frontend/src/components/EncryptedChat.js

```diff
@@ -52,155 +54,143 @@ class Chat extends Component {
-                  plainTextMessage: decryptedData.decryptedMessage,
-                  validSignature: decryptedData.validSignature,
-                  userNick: dataFromServer.user_nick,
-                  time: dataFromServer.time,
-                },
-              ].sort(function (a, b) {
-                // order the message array by their index (increasing)
-                return a.index - b.index;
-              }),
-            })),
-          );
+          console.log('PEER PUBKEY HAS CHANGED');
         }
+        console.log('PEER PUBKEY RECEIVED!!');
+        this.setState({ peer_pub_key: dataFromServer.message });
 
-        // We allow plaintext communication. The user must write # to start
-        // If we receive an plaintext message
-        else if (dataFromServer.message.substring(0, 1) == '#') {
-          console.log('Got plaintext message', dataFromServer.message);
+        // After receiving the peer pubkey we ask the server for the historic messages if any
+        this.state.send({
+          message: `-----SERVE HISTORY-----`,
+          nick: this.props.ur_nick,
+        });
+      }
+
+      // If we receive an encrypted message
+      else if (
+        dataFromServer.message.substring(0, 27) == `-----BEGIN PGP MESSAGE-----` &&
+        dataFromServer.index > this.state.latestIndex
+      ) {
+        decryptMessage(
+          dataFromServer.message.split('\\').join('\n'),
+          dataFromServer.user_nick == this.props.ur_nick
+            ? this.state.own_pub_key
+            : this.state.peer_pub_key,
+          this.state.own_enc_priv_key,
+          this.state.token,
+        ).then((decryptedData) =>
           this.setState((state) => ({
             scrollNow: true,
+            waitingEcho:
+              this.state.waitingEcho == true
+                ? decryptedData.decryptedMessage != this.state.lastSent
+                : false,
+            lastSent:
+              decryptedData.decryptedMessage == this.state.lastSent
+                ? '----BLANK----'
+                : this.state.lastSent,
+            latestIndex:
+              dataFromServer.index > this.state.latestIndex
+                ? dataFromServer.index
+                : this.state.latestIndex,
             messages: [
               ...state.messages,
               {
-                index: this.state.latestIndex + 0.001,
-                encryptedMessage: dataFromServer.message,
-                plainTextMessage: dataFromServer.message,
-                validSignature: false,
+                index: dataFromServer.index,
+                encryptedMessage: dataFromServer.message.split('\\').join('\n'),
+                plainTextMessage: decryptedData.decryptedMessage,
+                validSignature: decryptedData.validSignature,
                 userNick: dataFromServer.user_nick,
-                time: new Date().toString(),
+                time: dataFromServer.time,
```

Comment:
I just moved code to a function, github got crazy with the changes to display

Reply:
I'm certainly having a hard time getting my head around what has changed :smile: . On my tests the Chat does not work. Although the console logs "connected!", the `state.connected` remains `false` . I am sorry I am not able to debug this further at the moment.

Reply:
I'm certainly having a hard time getting my head around what has changed smile . On my tests the Chat does not work. Although the console logs "connected!", the state.connected remains false . I am sorry I am not able to debug this further at the moment.